opal/core/episodes.py

```diff
@@ -63,20 +65,58 @@ def has_stage(self, stage):
         """
         return stage in self.get_stages()
 
-    def set_stage(self, stage, user, data):
+    @transaction.atomic
+    def set_stage(self, stage_value, user, data):
         """
         Setter for Episode.stage
 
         Validates that the stage being set is appropriate for the category
         and raises ValueError if not.
         """
-        if not self.has_stage(stage):
-            if stage is not None:
+        if not self.has_stage(stage_value):
+            if stage_value is not None:
                 msg = "Can't set stage to {0} for {1} Episode".format(
-                    stage, self.display_name
+                    stage_value, self.display_name
                 )
                 raise ValueError(msg)
-        self.episode.stage = stage
+
+        current_stage = self.episode.stage_set.filter(stopped=None).last()
+
+        # if the stage hasn't changed, don't do anything
+        if current_stage and current_stage.value == stage_value:
+            return
+
+        # some models won't have stages yet, but if an episode category
+        # has been migrated to use stages,
+        # then it should once an episode has a stage it should
+        # always have a stage.
+
+        if current_stage and not stage_value:
+            raise ValueError(
+                "A stage cannot be removed after one has been set"
```

Comment:
Can we make the error message explain what has actually happened ? 
e.g. I'm not 'removing' a stage, I'm passing an invalid value, which we know about...